compiler/noirc_evaluator/src/brillig/brillig_gen/brillig_block.rs

```diff
@@ -1705,6 +1690,115 @@ impl<'block> BrilligBlock<'block> {
         }
     }
 
+    fn initialize_constant_array(
+        &mut self,
+        data: &im::Vector<ValueId>,
+        typ: &Type,
+        dfg: &DataFlowGraph,
+        pointer: MemoryAddress,
+    ) {
+        if data.is_empty() {
+            return;
+        }
+        let item_types: &Vec<_> = match typ {
+            Type::Array(item_types, _) => item_types.borrow(),
+            Type::Slice(item_types) => item_types.borrow(),
+            _ => unreachable!("ICE: Array constant is not of array or slice type {typ:?}"),
+        };
+
+        // Find out if we are repeating the same item over and over
+        let mut data_as_unique_items = HashSet::default();
+        for item_index in (0..data.len()).step_by(item_types.len()) {
+            let item: Vec<_> = (0..item_types.len()).map(|i| data[item_index + i]).collect();
+            data_as_unique_items.insert(item);
+        }
+
+        // If all the items are single address, and all have the same initial value, we can initialize the array in a runtime loop.
+        // Since the cost in instructions for a runtime loop is in the order of magnitude of 10, we only do this if the item_count is bigger than that.
+        let item_count = data.len() / item_types.len();
+
+        if item_count > 10
+            && data_as_unique_items.len() == 1
+            && item_types.iter().all(|typ| matches!(typ, Type::Numeric(_)))
+        {
```

Comment:
This function is quite large - maybe you can break up the then and else cases of this if into separate functions